### ISSUE TITLE
bad advice in the wiki

### ISSUE THREAD

**Issue by xendk** [minad/corfu#257]
Well, I figured out why I didn't get any candidates at all. Turns out that following your https://github.com/minad/corfu/wiki#configuring-corfu-for-lsp-mode breaks it for lsp-mode in it's current version.

In particular:

```
    (setf (alist-get 'styles (alist-get 'lsp-capf completion-category-defaults))
          '(flex))) ;; Configure flex
```

Will overwrite the `lsp-passthrough` style that lsp-completion sets. The style in turn is `(completion-basic-try-completion
              lsp-completion-passthrough-all-completions
              "Passthrough completion.")`

So basic completion plus:

```
(defun lsp-completion-passthrough-all-completions (_string table pred _point)
  "Like `completion-basic-all-completions' but have prefix ignored.
TABLE PRED"
  (completion-basic-all-completions "" table pred 0))
```

"What on earth is that?" you might rightfully ask, and I assume it's an attempt to accommodate both sides of the equation (what the lsp server returns and what capf works with).

I'm guessing here, but lsp seems to be using the `label` field from the reply, which is `my_var` when I complete on `$my`, while the `insertText` of the reply (`$my_var` in this case) is what lsp ends up inserting if selecting the `my_var` candidate. 

But I believe (correct me if I'm wrong) that all the completion functions are WYSIWYG, so, for instance, `completion-basic-all-completions` would throw away `my_var` because it doesn't match the `$my` prefix (and orderless did too). Hence, the prefix dropping wrapper. It might have other interesting effects with other language servers, as the protocol allows the server to return a completion with any label, a replacement that goes outside the bounds of the prefix of the completion, and additional text edits for other parts of the file (which lsp-mode actually support).

Obviously, this should be fixed in lsp-mode, but I fear I'm a bit in over my head, but I have a reasonable simple idea that I'll try out at a better time of day.

Anyway, I thought, despite your gruff locking of the previous issue, that you might be interested in knowing of this incompatibility.

**Comment by xendk** [comment]
But there's nothing to fix, so I'll close this again.

**Comment by minad** [comment]
Actually I think the bug is in the lsp server (not lsp-mode), since the label field itself misses the $.

**Comment by xendk** [comment]
No, it's a feature. Allows for having a popup that only contains the name of the variable (the icon selected for the type making it obvious that it's a variable). There's additional `insertText`, `filterText` and `sortText` for what's inserted, filtering and sorting. But capf doesn't differenciate between display text and filter text, so my "simple idea" is to make lsp use filterText instead. Has other consequences though.

> rerequest the candidates on every key press from the lsp server

Ugh, doesn't sound like that'll perform well. Might be worth a try until I get hacking on lsp-mode.

**Comment by xendk** [comment]
So, turns out that the `lsp-passthrough` thing was implemented as a response to an issue crated by you https://github.com/emacs-lsp/lsp-mode/issues/2970, and the fix was reviewed by you https://github.com/emacs-lsp/lsp-mode/pull/2975.

So to recap: The "configuration issue" you didn't want to help debug, was caused by following your instructions on configuration, which didn't work with the changes you requested in lsp-mode.

After reading through the issue and PR I get the feeling that just waltzing in there with "hey, I got this simple fix that might degrade the completions slightly but make it work in more cases" is going to go down too well.

I think my best bet would be to add another `lsp-completion-provider` option that implements a more "vanilla-capf" compatibility mode that works better with corfu.
